### PR TITLE
Support pageOrder in page setup for Xlsx/Xls Readers/Writers, and implement basic page setup support for other Readers/Writers 

### PR DESCRIPTION
**Pull request by MarkBaker**
This is:

```
- [ ] a bugfix
- [X] a new feature
```

Checklist:

- [X] Changes are covered by unit tests
- [X] Code style is respected
- [X] PR Description message explains **why** the change is made (see https://github.com/erlang/otp/wiki/Writing-good-commit-messages) and what has been added
- [X] CHANGELOG.md contains a short summary of the change
- [ ] Documentation is updated as necessary

### Why this change is needed?

Simple resolution for Issue [#1528](https://github.com/PHPOffice/PhpSpreadsheet/issues/1528)

Page Order implemented for:
- Xlsx Reader and Writer
- Xls Reader and Writer

Basic page setup also implemented for:
- Ods Reader
  - Orientation
  - Scale
  - Horizontal/Vertical Centering
  - Page Order
  - Page Margins
- Gnumeric Reader
  - Orientation
  - Scale
  - Horizontal/Vertical Centering
  - Page Order
  - Page Margins
- Xml Reader
  - Orientation
  - Scale
  - Horizontal/Vertical Centering
  - Page Order
  - Page Margins

### CODE REVIEW

**Comment by PowerKiKi** [comment]
@MarkBaker I see you regularly commit code style changes because of PHP-Cs-Fixer. Locally I use a git hook to automatically apply code style before committing. Avoiding the hassle of doing that manually. I suggest you set up something similar, so you can focus on more important things.

My setup is in `.git/hooks/pre-commit`:

```sh
#!/bin/bash

pass=true

files=$(git diff --cached --name-only --diff-filter=ACMR | grep -E '\.(php|phtml)$')
if [ "$files" != "" ]; then

    # Run php syntax check before commit
    while read -r file; do
        php -l "$file"
        if [ $? -ne 0 ]; then
            pass=false
        fi
    done <<< "$files"

    # Run php-cs-fixer validation before commit
    echo "$files" | xargs ./vendor/bin/php-cs-fixer fix --diff --config .php_cs.dist
    if [ $? -ne 0 ]; then
        pass=false
    fi

    # Automatically add files that may have been fixed by php-cs-fixer
    echo "$files" | xargs git add
fi

if $pass; then
    exit 0
else
    echo ""
    echo "PRE-COMMIT HOOK FAILED:"
    echo "Code style validation failed. Please fix errors and try committing again."
    exit 1
fi
```

File: src/PhpSpreadsheet/Reader/Gnumeric/PageSetup.php [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+     * @var Spreadsheet
+     */
+    private $spreadsheet;
+
+    /**
+     * @var string
+     */
+    private $gnm;
+
+    public function __construct(Spreadsheet $spreadsheet, $gnm)
```
Comment (PowerKiKi) [inline src/PhpSpreadsheet/Reader/Gnumeric/PageSetup.php:22]:
```suggestion
    public function __construct(Spreadsheet $spreadsheet, string $gnm)
```

File: src/PhpSpreadsheet/Worksheet/PageMargins.php [hunk 2]
```diff
@@ -211,4 +211,34 @@ public function __clone()
             }
         }
     }
+
+    public static function fromCentimeters($value): float
```
Comment (PowerKiKi) [inline src/PhpSpreadsheet/Worksheet/PageMargins.php:215]:
```suggestion
    public static function fromCentimeters(float $value): float
```

File: src/PhpSpreadsheet/Worksheet/PageMargins.php [hunk 3]
```diff
@@ -211,4 +211,34 @@ public function __clone()
...
+    {
+        return $value / 2.54;
+    }
+
+    public static function toCentimeters($value): float
```
Comment (PowerKiKi) [inline src/PhpSpreadsheet/Worksheet/PageMargins.php:220]:
```suggestion
    public static function toCentimeters(float $value): float
```

File: src/PhpSpreadsheet/Worksheet/PageMargins.php [hunk 4]
```diff
@@ -211,4 +211,34 @@ public function __clone()
...
+    {
+        return $value * 2.54;
+    }
+
+    public static function fromMillimeters($value): float
```
Comment (PowerKiKi) [inline src/PhpSpreadsheet/Worksheet/PageMargins.php:225]:
```suggestion
    public static function fromMillimeters(float $value): float
```

File: src/PhpSpreadsheet/Worksheet/PageMargins.php [hunk 5]
```diff
@@ -211,4 +211,34 @@ public function __clone()
...
+    {
+        return $value / 25.4;
+    }
+
+    public static function toMillimeters($value): float
```
Comment (PowerKiKi) [inline src/PhpSpreadsheet/Worksheet/PageMargins.php:230]:
```suggestion
    public static function toMillimeters(float $value): float
```

File: src/PhpSpreadsheet/Worksheet/PageMargins.php [hunk 6]
```diff
@@ -211,4 +211,34 @@ public function __clone()
...
+    {
+        return $value * 25.4;
+    }
+
+    public static function fromPoints($value): float
```
Comment (PowerKiKi) [inline src/PhpSpreadsheet/Worksheet/PageMargins.php:235]:
```suggestion
    public static function fromPoints(float $value): float
```

File: src/PhpSpreadsheet/Worksheet/PageMargins.php [hunk 7]
```diff
@@ -211,4 +211,34 @@ public function __clone()
...
+    {
+        return $value / 72;
+    }
+
+    public static function toPoints($value): float
```
Comment (PowerKiKi) [inline src/PhpSpreadsheet/Worksheet/PageMargins.php:240]:
```suggestion
    public static function toPoints(float $value): float
```

File: tests/PhpSpreadsheetTests/Reader/Gnumeric/PageSetupTest.php [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+                    $expectedResult,
+                    $actualResult,
+                    self::MARGIN_PRECISION,
+                    "Failed assertion for Worksheet '{$worksheet->getTitle()}' {$test} margin"
+                );
+            }
+        }
+    }
+
+    public function pageSetupAssertions()
```
Comment (PowerKiKi) [inline tests/PhpSpreadsheetTests/Reader/Gnumeric/PageSetupTest.php:71]:
```suggestion
    public function pageSetupAssertions(): array
```

File: tests/PhpSpreadsheetTests/Reader/Gnumeric/PageSetupTest.php [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+                'orientation' => PageSetup::ORIENTATION_PORTRAIT,
+                'scale' => 100,
+                'horizontalCentered' => false,
+                'verticalCentered' => false,
+                'pageOrder' => PageSetup::PAGEORDER_DOWN_THEN_OVER,
+            ],
+        ];
+    }
+
+    public function pageMarginAssertions()
```
Comment (PowerKiKi) [inline tests/PhpSpreadsheetTests/Reader/Gnumeric/PageSetupTest.php:106]:
```suggestion
    public function pageMarginAssertions(): array
```

File: tests/PhpSpreadsheetTests/Reader/Ods/PageSetupTest.php [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+                    $expectedResult,
+                    $actualResult,
+                    self::MARGIN_PRECISION,
+                    "Failed assertion for Worksheet '{$worksheet->getTitle()}' {$test} margin"
+                );
+            }
+        }
+    }
+
+    public function pageSetupAssertions()
```
Comment (PowerKiKi) [inline tests/PhpSpreadsheetTests/Reader/Ods/PageSetupTest.php:73]:
```suggestion
    public function pageSetupAssertions(): array
```

File: tests/PhpSpreadsheetTests/Reader/Ods/PageSetupTest.php [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+                'orientation' => PageSetup::ORIENTATION_DEFAULT,
+                'scale' => 100,
+                'horizontalCentered' => false,
+                'verticalCentered' => false,
+                'pageOrder' => PageSetup::PAGEORDER_DOWN_THEN_OVER,
+            ],
+        ];
+    }
+
+    public function pageMarginAssertions()
```
Comment (PowerKiKi) [inline tests/PhpSpreadsheetTests/Reader/Ods/PageSetupTest.php:108]:
```suggestion
    public function pageMarginAssertions(): array
```

File: tests/PhpSpreadsheetTests/Reader/Xls/PageSetupTest.php [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+                    $expectedResult,
+                    $actualResult,
+                    self::MARGIN_PRECISION,
+                    "Failed assertion for Worksheet '{$worksheet->getTitle()}' {$test} margin"
+                );
+            }
+        }
+    }
+
+    public function pageSetupAssertions()
```
Comment (PowerKiKi) [inline tests/PhpSpreadsheetTests/Reader/Xls/PageSetupTest.php:73]:
```suggestion
    public function pageSetupAssertions(): array
```

File: tests/PhpSpreadsheetTests/Reader/Xls/PageSetupTest.php [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+                'orientation' => PageSetup::ORIENTATION_DEFAULT,
+                'scale' => 100,
+                'horizontalCentered' => false,
+                'verticalCentered' => false,
+                'pageOrder' => PageSetup::PAGEORDER_DOWN_THEN_OVER,
+            ],
+        ];
+    }
+
+    public function pageMarginAssertions()
```
Comment (PowerKiKi) [inline tests/PhpSpreadsheetTests/Reader/Xls/PageSetupTest.php:108]:
```suggestion
    public function pageMarginAssertions(): array
```

File: tests/PhpSpreadsheetTests/Reader/Xlsx/PageSetupTest.php [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+                    $expectedResult,
+                    $actualResult,
+                    self::MARGIN_PRECISION,
+                    "Failed assertion for Worksheet '{$worksheet->getTitle()}' {$test} margin"
+                );
+            }
+        }
+    }
+
+    public function pageSetupAssertions()
```
Comment (PowerKiKi) [inline tests/PhpSpreadsheetTests/Reader/Xlsx/PageSetupTest.php:73]:
```suggestion
    public function pageSetupAssertions(): array
```

File: tests/PhpSpreadsheetTests/Reader/Xlsx/PageSetupTest.php [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+                'orientation' => PageSetup::ORIENTATION_DEFAULT,
+                'scale' => 100,
+                'horizontalCentered' => false,
+                'verticalCentered' => false,
+                'pageOrder' => PageSetup::PAGEORDER_DOWN_THEN_OVER,
+            ],
+        ];
+    }
+
+    public function pageMarginAssertions()
```
Comment (PowerKiKi) [inline tests/PhpSpreadsheetTests/Reader/Xlsx/PageSetupTest.php:108]:
```suggestion
    public function pageMarginAssertions(): array
```

File: tests/PhpSpreadsheetTests/Reader/Xml/PageSetupTest.php [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+                    $expectedResult,
+                    $actualResult,
+                    self::MARGIN_PRECISION,
+                    "Failed assertion for Worksheet '{$worksheet->getTitle()}' {$test} margin"
+                );
+            }
+        }
+    }
+
+    public function pageSetupAssertions()
```
Comment (PowerKiKi) [inline tests/PhpSpreadsheetTests/Reader/Xml/PageSetupTest.php:73]:
```suggestion
    public function pageSetupAssertions(): array
```

File: tests/PhpSpreadsheetTests/Reader/Xml/PageSetupTest.php [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+                'orientation' => PageSetup::ORIENTATION_DEFAULT,
+                'scale' => 100,
+                'horizontalCentered' => false,
+                'verticalCentered' => false,
+                'pageOrder' => PageSetup::PAGEORDER_DOWN_THEN_OVER,
+            ],
+        ];
+    }
+
+    public function pageMarginAssertions()
```
Comment (PowerKiKi) [inline tests/PhpSpreadsheetTests/Reader/Xml/PageSetupTest.php:108]:
```suggestion
    public function pageMarginAssertions(): array
```

File: tests/PhpSpreadsheetTests/Worksheet/PageMarginsTest.php [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+    /**
+     * @dataProvider providerMillimetersAndInches
+     */
+    public function testPointsToMillimeters(float $expectedResult, float $value): void
+    {
+        $actualResult = PageMargins::toMillimeters($value);
+        self::assertSame($expectedResult, $actualResult);
+    }
+
+    public function providerPointsAndInches()
```
Comment (PowerKiKi) [inline tests/PhpSpreadsheetTests/Worksheet/PageMarginsTest.php:64]:
```suggestion
    public function providerPointsAndInches(): array
```

File: tests/PhpSpreadsheetTests/Worksheet/PageMarginsTest.php [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        return [
+            [36, 0.5],
+            [72, 1.0],
+            [90, 1.25],
+            [144, 2.0],
+        ];
+    }
+
+    public function providerCentimetersAndInches()
```
Comment (PowerKiKi) [inline tests/PhpSpreadsheetTests/Worksheet/PageMarginsTest.php:74]:
```suggestion
    public function providerCentimetersAndInches(): array
```

**Comment by MarkBaker** [comment]
Again, I'm disinclined to use `AbstractFunctional` because it only really proves that we can read a file that we've generated ourselves, and not one generated by MS Excel or other external spreadsheet program. I think that as long as we keep the binaries small, and then they are justified for testing Readers

**Comment by PowerKiKi** [comment]
Fair enough, then it's good to merge for me.

**Pull request merged**